internal/pkg/gateway/api.go

```diff
@@ -79,32 +93,45 @@ func (gs *Server) Endorse(ctx context.Context, proposedTransaction *gp.ProposedT
 	// send to all the endorsers
 	for _, e := range endorsers {
 		wg.Add(1)
-		go func(endorser *endorser) {
+		go func(e *endorser) {
 			defer wg.Done()
-			response, err := endorser.client.ProcessProposal(ctx, signedProposal)
-			responseCh <- &endorserResponse{pr: response, err: err}
+			response, err := e.client.ProcessProposal(ctx, signedProposal)
+			var detail *gp.EndpointError
+			if err != nil {
+				detail = &gp.EndpointError{Address: e.address, MspId: e.mspid, Message: err.Error()}
+			}
+			if response.Response.Status < 200 || response.Response.Status >= 400 {
+				detail = &gp.EndpointError{Address: e.address, MspId: e.mspid, Message: fmt.Sprintf("error %d, %s", response.Response.Status, response.Response.Message)}
+			}
```

Comment:
This block looks quite suspect to me. I think the tests need a bit more focus.

Reply:
Which bit is suspect?  If either the grpc call to the remote endorser failed (i.e. returned non-nil error) or the chaincode function returned a non-success status code, then it adds the error detail to the response.
If any of the endorsers produce an error like this, then the whole Endorse() function will abort and return the details to the caller.  It wouldn't be able to continue because the subsequent call to `protoutil.CreateTx` will perform the same check and fail.
I think I'm testing both of these scenarios in the unit tests.
Or have I misunderstood your comment?

Reply:
Take a look a the client.

https://github.com/hyperledger/fabric-protos-go/blob/e08edaab0493270bf5b386b3d458036fad2b0000/peer/peer.pb.go#L68-L75